examples/server/server.cpp

```diff
@@ -1923,6 +1954,58 @@ struct server_context {
         return slot.has_next_token; // continue
     }
 
+    void populate_token_probs(const server_slot & slot, completion_token_output & result, bool post_sampling, bool special, int idx) {
+        size_t n_probs = slot.params.sampling.n_probs;
+        int n_vocab = llama_n_vocab(llama_get_model(ctx));
+        if (post_sampling) {
+            const auto * cur_p = common_sampler_get_candidates(slot.smpl);
+            const size_t max_probs = cur_p->size;
+
+            bool found_sampled_tok = false;
+            result.probs.reserve(max_probs);
+            for (size_t i = 0; i < max_probs; i++) {
+                // set probability for sampled token
+                if (cur_p->data[i].id == result.tok) {
+                    found_sampled_tok = true;
+                    result.prob = cur_p->data[i].p;
+                }
+                // set probability for top n_probs tokens
+                result.probs.push_back({
+                    cur_p->data[i].id,
+                    common_detokenize(ctx, {cur_p->data[i].id}, special),
+                    cur_p->data[i].p
+                });
+                // break if we have all the necessary data
+                if (result.probs.size() == n_probs && found_sampled_tok) {
```

Comment:
This seems incorrect - if the sampled token is from the tail of the distribution, we will never break here.

Reply:
Hmm yeah it's quite messy to do every thing in one loop. I think it's better to split this into 2 different loops, and then optimize the one that take the prob of sampled token later on (in another PR)